ggml/src/ggml-sycl/dpct/math.hpp

```diff
@@ -0,0 +1,237 @@
+//
+// MIT license
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: MIT
+//
+
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+
+/***************************************************************************
+ *
+ *  Copyright (C) Codeplay Software Ltd.
+ *
+ *  Part of the LLVM Project, under the Apache License v2.0 with LLVM
+ *  Exceptions. See https://llvm.org/LICENSE.txt for license information.
+ *  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ **************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+#include <type_traits>
+
+#include <sycl/sycl.hpp>
+
+namespace dpct {
+
+namespace detail {
+
+template <typename VecT, class BinaryOperation, class = void>
+class vectorized_binary {
+  public:
+    inline VecT operator()(VecT a, VecT b, const BinaryOperation binary_op) {
+        VecT v4;
+        for (size_t i = 0; i < v4.size(); ++i) {
+            v4[i] = binary_op(a[i], b[i]);
+        }
+        return v4;
+    }
+};
+
+template <typename VecT, class BinaryOperation>
+class vectorized_binary<
+    VecT, BinaryOperation,
+    std::void_t<std::invoke_result_t<BinaryOperation, VecT, VecT>>> {
+  public:
+    inline VecT operator()(VecT a, VecT b, const BinaryOperation binary_op) {
+        return binary_op(a, b).template as<VecT>();
+    }
+};
+
+} // namespace detail
+
+template <typename T> sycl::vec<T, 4> extract_and_sign_or_zero_extend4(T val) {
+    return sycl::vec<T, 1>(val)
+        .template as<sycl::vec<
+            std::conditional_t<std::is_signed_v<T>, int8_t, uint8_t>, 4>>()
+        .template convert<T>();
+}
+
+template <typename T1, typename T2>
+using dot_product_acc_t =
+    std::conditional_t<std::is_unsigned_v<T1> && std::is_unsigned_v<T2>,
+                       uint32_t, int32_t>;
+
+template <typename T1, typename T2, typename T3>
+inline auto dp4a(T1 a, T2 b, T3 c) {
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__) &&                     \
+    defined(__SYCL_CUDA_ARCH__) && __SYCL_CUDA_ARCH__ >= 610
+    dot_product_acc_t<T1, T2> res;
+    if constexpr (std::is_same_v<dot_product_acc_t<T1, T2>, uint32_t>) {
+        asm volatile("dp4a.u32.u32 %0, %1, %2, %3;"
+                     : "=r"(res)
+                     : "r"(a), "r"(b), "r"(c));
+    } else {
+        asm volatile("dp4a.s32.s32 %0, %1, %2, %3;"
+                     : "=r"(res)
+                     : "r"(a), "r"(b), "r"(c));
+    }
```

Comment:
we'd better limit it within SYCL range

Reply:
SYCL backend should be pure SYCL

1. If you want extra optimization on cuda device, they should also be based on SYCL and not surpassed SYCL capabilities, thus no cuda assembly introduced
2. If you can contribute cuda specific optimization to SYCL , like intel_ext in SYCL. llama.cpp SYCL backend can apply it

Reply:
@Alcpz  maybe we can use dpct asm instructions for this and test for ptx ?